packages/core/src/lib/markdown-it/highlight/HighlightRule.js

```diff
@@ -11,8 +9,36 @@ class HighlightRule {
     this.ruleComponents = ruleComponents;
   }
 
+  static parseAllRules(allRules, lineOffset, tokenContent) {
+    const highlightLines = this.splitByChar(allRules, ',');
+    const strArray = tokenContent.split('\n');
+    return highlightLines
+      .map(ruleStr => HighlightRule.parseRule(ruleStr, lineOffset, strArray))
+      .filter(rule => rule); // discards invalid rules
+  }
+
+  static splitByChar(allRules, splitter) {
+    const highlightRules = [];
+    let isWithinQuote = false;
+    let currentPosition = 0;
+    for (let i = 0; i < allRules.length; i += 1) {
```

Comment:
On that note, I think we can consider using a `split` method instead of a `for` loop. For instance, we can initially split the rules based on unescaped quotation mark using something like `.split(\(?<!\\)\'\)`, and then further split the substrings using the `splitter` given to the `splitByChar` method (not sure if this will fully remove the need for `currentPosition` but looks like it). What do you think @yucheng11122017?

Reply:
Hmmm I'm not too sure how that will work. The point of this function is to split by the splitter except for when the splitter is within a quote Eg. ```','```
If I split using the unescaped quotes, then split by the splitter won't it give broken outputs?
Eg. ```['hi'], 1['sd']``` will give ```'[', 'hi', ']', '1[', 'sd', ']'```

Am I interpreting you correctly?

Reply:
Given that the regex I mentioned above will not consider `\'`, I was thinking that we could use this to filter out the outputs that are not within the quotation marks, and only split the parts that were inside the quotation marks by the splitter. My reasoning was that if we only split by `'` then the split should alternate between the segments that were inside the quotation marks and segments that were outside the quotation marks. E.g. `['hi'], 1['sd']` will result in `'[', 'hi', '], 1[', 'sd', ']'`. We can then split the alternating substrings of `'hi'` and `'sd'` using the `splitter`, etc.

Reply:
Hmm but the point of the function is to split the function according to the splitter, not the quotation though. Like it doesn't need the split according the quotation mark

Reply:
Ah, I see! Piecing it back together would indeed be a problem. In that case the current implementation should be better than the suggestion :sweat_smile:

Reply:
Alright thanks for all the suggestions though! @lhw-1